nautobot/dcim/migrations/0023_create_region_and_site_location_types_and_locations.py

```diff
@@ -0,0 +1,150 @@
+            location_type.parent = site_lt
+            location_type.save()
+        for location in Location.objects.filter(site__isnull=False):
+            location.parent = Location.objects.get(name=location.site.name, location_type=site_lt)
+            location.save()
+
+    # Reassign Site Models to Locations of Site LocationType
+    if Site.objects.exists():  # Iff Site instances exist
+        CircuitTermination = apps.get_model("circuits", "circuittermination")
+        Device = apps.get_model("dcim", "device")
+        PowerPanel = apps.get_model("dcim", "powerpanel")
+        RackGroup = apps.get_model("dcim", "rackgroup")
+        Rack = apps.get_model("dcim", "rack")
+        Prefix = apps.get_model("ipam", "prefix")
+        VLANGroup = apps.get_model("ipam", "vlangroup")
+        VLAN = apps.get_model("ipam", "vlan")
+        Cluster = apps.get_model("virtualization", "cluster")
+
+        for ct in CircuitTermination.objects.filter(location__isnull=True):
+            ct.location = Location.objects.get(name=ct.site.name, location_type=site_lt)
+            ct.save()
+        for device in Device.objects.filter(location__isnull=True):
+            device.location = Location.objects.get(name=device.site.name, location_type=site_lt)
+            device.save()
+        for powerpanel in PowerPanel.objects.filter(location__isnull=True):
+            powerpanel.location = Location.objects.get(name=powerpanel.site.name, location_type=site_lt)
+            powerpanel.save()
+        for rackgroup in RackGroup.objects.filter(location__isnull=True):
+            rackgroup.location = Location.objects.get(name=rackgroup.site.name, location_type=site_lt)
+            rackgroup.save()
+        for rack in Rack.objects.filter(location__isnull=True):
+            rack.location = Location.objects.get(name=rack.site.name, location_type=site_lt)
+            rack.save()
+        # Below models' site attribute is not required, so we need to check each instance if the site field is not null
+        # if so we reassign it to Site Location and if not we leave it alone
+        for prefix in Prefix.objects.filter(location__isnull=True, site__isnull=False):
+            prefix.location = Location.objects.get(name=prefix.site.name, location_type=site_lt)
+            prefix.save()
+        for vlangroup in VLANGroup.objects.filter(location__isnull=True, site__isnull=False):
+            vlangroup.location = Location.objects.get(name=vlangroup.site.name, location_type=site_lt)
+            vlangroup.save()
+        for vlan in VLAN.objects.filter(location__isnull=True, site__isnull=False):
+            vlan.location = Location.objects.get(name=vlan.site.name, location_type=site_lt)
+            vlan.save()
+        for cluster in Cluster.objects.filter(location__isnull=True, site__isnull=False):
+            cluster.location = Location.objects.get(name=cluster.site.name, location_type=site_lt)
+            cluster.save()
+
```

Comment:
Refactor suggestion, My function names might be bad 👀:
```python

def create_region_location(region_class, location_class, region_lt):
     """
    Creates location objects for each region instance in the region_class model.

    Args:
        region_class: The model class for the regions.
        location_class: The model class for the locations.
        region_location_type: The location type for the regions.
    """
    # Breadth First Query
    regions = region_class.objects.with_tree_fields().extra(order_by=["__tree.tree_depth", "__tree.tree_ordering"])
    locations_instances = []
    for region in regions:
        parent = (
            location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent else None
            )
        locations_instances.append(
            location_class(
                location_type=region_lt,
                name=region.name,
                description=region.description,
                parent=parent
            )
        )
    location_class.objects.bulk_create(locations_instances, batch_size=1000)
    

def create_site_location(
    site_class, 
    location_type_class, 
    location_class, 
    site_lt, 
    exclude__name, 
    has_parent_location=False, 
    region_lt=None,
    ):
    """
    Creates location objects for each site instance in the site_class model.

    Args:
        site_class: The model class for the sites.
        location_type_class: The model class for the location types.
        location_class: The model class for the locations.
        site_location_type: The location type for the sites.
        exclude_name: The name to exclude from the update query.
        has_parent_location: Whether the sites have a parent location or not.
        region_location_type: The location type for the regions, required if has_parent_location is True.
    """
    
    locations_instances = []
    for site in site_class.objects.all():
        extra_kwargs = {}
        if has_parent_location:
            extra_kwargs["parent"] = location_class.objects.get(
                location_type=region_lt,
                name=site.region.name if site.region else "Global Region"
            )
        locations_instances.append(
            location_class(
                name=site.name,
                location_type=site_lt,
                tenant=site.tenant,
                facility=site.facility,
                asn=site.asn,
                time_zone=site.time_zone,
                description=site.description,
                physical_address=site.physical_address,
                shipping_address=site.shipping_address,
                latitude=site.latitude,
                longitude=site.longitude,
                contact_name=site.contact_name,
                contact_phone=site.contact_phone,
                contact_email=site.contact_email,
                comments=site.comments,
                status=site.status,
                tags=site.tags,
            )
        )
    location_class.objects.bulk_create(locations_instances, batch_size=1000)
    location_class.objects.filter(site__isnull=False).update(parent=F('site__name'), location_type=site_lt)
    location_type_class.objects.filter(parent__isnull=True).exclude(name=exclude__name).update(parent=site_lt)
    

def create_region_and_site_location_types(apps, schema_editor):
    Region = apps.get_model("dcim", "region")
    Site = apps.get_model("dcim", "site")
    LocationType = apps.get_model("dcim", "locationtype")
    Location = apps.get_model("dcim", "location")

    # Region instances exist
    if Region.objects.exists():
        region_lt = LocationType.objects.create(name="Region", nestable=True)
        create_region_location(
            region_class=Region, 
            location_type_class=LocationType, 
            location_class=Location,
            region_lt=region_lt,
        )
        if Site.objects.exists():  # Both Site and Region instances exist
            parent_location = LocationType.objects.get(name="Region")
            site_lt = LocationType.objects.create(name="Site", parent=parent_location)
            if Site.objects.filter(region__isnull=True).exists():
                Location.objects.create(
                    location_type=region_lt,
                    name="Global Region",
                    description="Parent Location of Region LocationType for all sites that "
                    "did not have a region attribute set before the migration",
                )
            create_site_location(
                location_class=Location,
                location_type_class=LocationType,
                site_class=Site,
                site_lt=site_lt,
                exclude__name="Region",
                has_parent_location=True,
                region_lt=region_lt,
            )


    elif Site.objects.exists():  # Only Site instances exist, we make Site the top level LocationType
        site_lt = LocationType.objects.create(name="Site")
        create_site_location(
            site_class=Region, 
            location_type_class=LocationType, 
            location_class=Location,
            site_lt=site_lt,
            exclude__name="Site",
        )

```

Reply:
I second this. My feedback was going to be that the existing function is hard to follow because it's so nested and also results in undesirable duplication of code.

Reply:
Just some modifications I have to make  with the code:
```
for region in regions:
        parent = (
            location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent else None
          )
```
`location_class.objects.get(name=region.parent.name, location_type=region_lt)` will always throw a ` no matching query exist` error because the Region Location Type locations are created until later. 
So I changed it back to creating one Region Location at a time.
```
regions = region_class.objects.with_tree_fields().extra(order_by=["__tree.tree_depth", "__tree.tree_ordering"])
for region in regions:
        location_class.objects.create(
            location_type=region_lt,
            name=region.name,
            description=region.description,
            parent=location_class.objects.get(name=region.parent.name, location_type=region_lt)
            if region.parent
            else None,
       )
```

`location_class.objects.filter(site__isnull=False).update(parent=F('site__name'), location_type=site_lt)` This line will throw an error below:
<img width="1043" alt="Screen Shot 2023-01-04 at 2 46 57 PM" src="https://user-images.githubusercontent.com/46973263/210645328-9f6d9610-ab00-44b4-86e7-fe8123fa321a.png">
So I have to change it back to
```
for location in location_class.objects.filter(site__isnull=False):
        location.parent = location_class.objects.get(location_type=site_lt, name=location.site.name)
        location.save()
```